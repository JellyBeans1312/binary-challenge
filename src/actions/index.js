export const addEvents = (events) => ({
  type: 'ADD_EVENTS',
  events
});