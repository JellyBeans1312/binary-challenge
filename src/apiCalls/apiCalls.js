export const getEvents = async () => {
  try {
    const options = { 
      method: 'GET',
      headers: {
        Authorization: 'Bearer qJhnqW5Rsc36m6cUYXZyCaV2BaFNB0QoGpr4fmk_',
        Accept: 'application/json',
        'Content-Type': 'application/json'
      }
    }

    const response = await fetch('https://api.predicthq.com/v1/events/', options)
    const result = await response.json()
    return result.results
  } catch(error) {
    console.log(error);
  }
}

export const searchEvent = async (searchParam) => {
  try {
    const options = {
      method: 'GET',
      headers: {
        Authorization: 'Bearer qJhnqW5Rsc36m6cUYXZyCaV2BaFNB0QoGpr4fmk_',
        Accept: 'application/json',
        'Content-Type': 'application/json'
      }
    }
    
    const response = await fetch(`https://api.predicthq.com/v1/events/?q=${searchParam}/`, options)
    const result = await response.json()
    return result.results

  } catch(error) {
    throw new Error(error)
  }
}

export const getAddress = async (lat, long) => {
  try {
    const response = await fetch(`https://maps.googleapis.com/maps/api/geocode/json?latlng=${lat},${long}&key=AIzaSyDPPlNyjATG8sQnfsdg45ln7kUoyFH24m0`)
    const result = await response.json();
    return result.results.map(result => {
      return result.formatted_address
    });
  } catch(error) {
    console.log(error)
  }
}