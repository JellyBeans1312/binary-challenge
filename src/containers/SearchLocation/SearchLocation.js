import React, { Component } from 'react';
import { getLocation, getEventsByLocation } from '../../apiCalls/apiCalls';
import { addEvents } from '../../actions/'
import { connect } from 'react-redux';

export class SearchLocation extends Component {
  constructor() {
    super();
    this.state = {
      city: '',
      state: ''
    }
  }

  onChange = (e) => {
    const name = e.target.name
    this.setState({ [name] : e.target.value })
  }

  searchEventByLocation = async (e) => {
    const { city, state } = this.state
    e.preventDefault()
    try {
      const location = await getLocation(city, state)
      const searchResults = await getEventsByLocation(location.geometry.location.lat, location.geometry.location.lng)
      this.props.addEvents(searchResults) 
    } catch(error) {
      this.setState({ error: 'There was a problem getting your search results'})
    }
  }

  render() {
    return (
      <form onSubmit={this.searchEventByLocation}>
        <input
          type='text'
          placeholder='Enter a city'
          value={this.state.city}
          onChange={this.onChange}
          name='city'
        />
        <input
          type='text'
          placeholder='Enter a state (CO format)'
          value={this.state.state}
          onChange={this.onChange}
          name='state'
        />
        <button>Search by location</button>
      </form>
    )
  }
}

export const mapDispatchToProps = dispatch => ({
  addEvents: events => dispatch(addEvents(events))
})

export default connect(null, mapDispatchToProps)(SearchLocation)