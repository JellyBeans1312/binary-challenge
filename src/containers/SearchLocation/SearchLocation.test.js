import React from 'react';
import { shallow } from 'enzyme';
import { SearchLocation, mapDispatchToProps } from './SearchLocation';
import { getLocation, getEventsByLocation }  from '../../apiCalls/apiCalls';
import { addEvents } from '../../actions'
jest.mock('../../apiCalls/apiCalls')

describe('SearchLocation', () => {
  let wrapper;
  let mockEvent;

  const mockAddEvents = jest.fn()

  const mockEvents = [
    { title: "something", category: 'concert', location: [{lat: 23, long: 234}] }
  ];

  beforeEach(() => {
    mockEvent = {preventDefault: jest.fn()}
    wrapper = shallow(
      <SearchLocation
        addEvents={mockAddEvents}
      />
    )
    getLocation.mockImplementation(() => {
      return Promise.resolve([{address: 'something 123 something else drive'}])
    });
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  });

  it('should call onChange when change happens', () => {
    const mockEvent = { target: { value: 'top down and im floating through the city', name: 'city'} }
    wrapper.find('input').at(0).simulate('change', mockEvent)
    expect(wrapper.state('city')).toEqual('top down and im floating through the city')
  });
  
  it('should call getLocation when searchEventByLocation is invoked', () => {
    wrapper.instance().searchEventByLocation(mockEvent)
    expect(getLocation).toHaveBeenCalled()
  });

  it('should set error in state if any of the fetches fail', async () => {
    getLocation.mockImplementationOnce(() => {
      throw new Error("Please Try Again");
    });

    wrapper.instance().searchEventByLocation(mockEvent);

    expect(wrapper.state("error")).toEqual("There was a problem getting your search results");
  });

  it('should call getEventsByLocation when searchEventByLocation is invoked', () => {
    getEventsByLocation.mockImplementation(() => {
      return Promise.resolve(mockEvents);
    });
    wrapper.instance().searchEventByLocation(mockEvent)
    expect(getEventsByLocation).toHaveBeenCalled()
  });

  it('should call add events when searchEventByLocation is invoked', () => {
  
    wrapper.instance().searchEventByLocation(mockEvent)
    expect(mockAddEvents).toHaveBeenCalled()
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with addEvents or something like that', () => {
      const mockDispatch = jest.fn()
      const actionToDispatch = addEvents(mockEvents)
      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.addEvents(mockEvents)

      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    });
  });
});