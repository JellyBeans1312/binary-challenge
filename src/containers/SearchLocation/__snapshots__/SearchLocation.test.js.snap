// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchLocation should match snapshot 1`] = `
<section>
  <form
    onSubmit={[Function]}
  >
    <input
      name="city"
      onChange={[Function]}
      placeholder="Enter a city"
      type="text"
      value=""
    />
    <input
      name="state"
      onChange={[Function]}
      placeholder="Enter a state (CO format)"
      type="text"
      value=""
    />
    <button>
      Search by location
    </button>
  </form>
  <Connect(EventContainer) />
</section>
`;
